contracts/token/LinearVestingWithPenalty.sol

```diff
@@ -415,37 +280,122 @@ contract LinearVestingWithPenalty is Initializable, ERC20Upgradeable, AccessCont
+    /**
+    * @dev Returns the list of unclaimed vesting schedule indexes for the beneficiary within a specified range and their respective releasable amounts.
+    * @param beneficiary The address of the beneficiary to check unclaimed schedules for.
+    * @param start The starting index of the range (inclusive).
+    * @param end The ending index of the range (exclusive).
+    * @return indexes An array of indexes for schedules with unclaimed tokens within the specified range.
+    * @return amounts An array of unclaimed amounts corresponding to each index within the specified range.
+    */
+    function getUnclaimedSchedulesInRange(
+        address beneficiary,
+        uint256 start,
+        uint256 end
+    ) external view returns (uint256[] memory indexes, uint256[] memory amounts) {
+        uint256 scheduleCount = vestingSchedules[beneficiary].length;
+        require(start < end && end <= scheduleCount, "Invalid range");
+
+        // Create temporary arrays with a fixed maximum size (end - start)
+        uint256[] memory tempIndexes = new uint256[](end - start);
+        uint256[] memory tempAmounts = new uint256[](end - start);
+        uint256 unclaimedCount = 0;
+
+        // Single loop to collect unclaimed schedules within the range
+        for (uint256 i = start; i < end; i++) {
+            VestingSchedule storage schedule = vestingSchedules[beneficiary][i];
+            uint256 releasable = _releasableAmount(schedule);
+
+            if (releasable > 0) {
+                tempIndexes[unclaimedCount] = i;
+                tempAmounts[unclaimedCount] = releasable;
+                unclaimedCount++;
+            }
         }
 
-        delete vestingSchedules[from];
+        // Create result arrays with exact size of unclaimed schedules
+        indexes = new uint256[](unclaimedCount);
+        amounts = new uint256[](unclaimedCount);
+
+        // Copy collected data to result arrays
+        for (uint256 j = 0; j < unclaimedCount; j++) {
+            indexes[j] = tempIndexes[j];
+            amounts[j] = tempAmounts[j];
+        }
+
+        return (indexes, amounts);
+    }
+
+
+    /**
+    * @dev Claims the entire releasable amount for specific vesting schedule indexes with an applied penalty.
+    * @param scheduleIndexes An array of indexes of the vesting schedules to claim from.
+    */
+    function claimSpecificSchedulesWithPenalty(uint256[] calldata scheduleIndexes) external {
+        if (vestingStartDate > block.timestamp) revert VestingHasNotStartedYet();
+
+        uint256 totalReleasable = 0;
+        uint256 totalPenaltyAmount = 0;
+
+        for (uint256 i = 0; i < scheduleIndexes.length; i++) {
+            uint256 index = scheduleIndexes[i];
+            if (index >= vestingSchedules[msg.sender].length) revert InvalidScheduleIndex();
+
+            VestingSchedule storage schedule = vestingSchedules[msg.sender][index];
+            uint256 releasable = _releasableAmount(schedule);
```

Comment:
I'm a bit confused here. `claimSpecificSchedulesWithPenalty` is supposed to allow "early" claims? And if user does this, he would get penalised, right?

If so, shouldn't this be something like:

`uint256 total = schedule.totalAmount - schedule.released; `

?

Reply:
fixed